### PR TITLE
File/Folder Reorg Changes with Backward Compatibility Support

### PR DESCRIPTION
**Pull request by arvindcheru**
File/Folder Reorg Changes
Backward Compatibility Support enabled by default

### CODE REVIEW

File: install [hunk 1]
```diff
@@ -111,6 +113,9 @@ while true; do
         --address-sanitizer)
             build_address_sanitizer=true
             shift ;;
+        --freorg-bkwdcomp)
+            build_freorg_bkwdcomp=true
```
Comment (TorreZuk) [inline install:117]: If off by default you need the option to turn it off as I described in the email

File: CMakeLists.txt [hunk 2]
```diff
@@ -108,6 +108,18 @@ option(BUILD_EXAMPLE "Build examples" OFF)
 option(ONLY_INSTALL "Only install" OFF)
 option(BUILD_CODE_COVERAGE "Build with code coverage enabled" OFF)
 
+# FOR HANDLING ENABLE/DISABLE OPTIONAL BACKWARD COMPATIBILITY for FILE/FOLDER REORG
+option(BUILD_FILE_REORG_BACKWARD_COMPATIBILITY "Build with file/folder reorg with backward compatibility enabled" OFF)
+if(BUILD_FILE_REORG_BACKWARD_COMPATIBILITY)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DFILE_REORG_BACKWARD_COMPATIBILITY")
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -DFILE_REORG_BACKWARD_COMPATIBILITY")
```
Comment (TorreZuk) [inline CMakeLists.txt:115]: These CMAKE_CXX_FLAGS and CMAKE_C_FLAGS may trigger rocm-checks warnings and why do you need them as I don't see you using those defines within the actual c/c++ files?

**Review by TorreZuk — COMMENTED** [review]
Some general comments leave for Stanley to review.

File: header.hpp.in [hunk 3]
```diff
@@ -0,0 +1,8 @@
+/* ************************************************************************
+ * Copyright 2019-2022 Advanced Micro Devices, Inc.
+ * ************************************************************************ */
+
+/* The Wrapper template Configuration & settings
+ */
+#warning "This file is deprecated and will be removed in future releases. Use the include path /opt/rocm-ver/include and prefix with rocprim."
```
Comment (TorreZuk) [inline header.hpp.in:7]: Sorry missed this should it be ..."prefix all includes with rocprim/"

File: install [hunk 4]
```diff
@@ -40,6 +41,7 @@ build_benchmark=false
 rocm_path=/opt/rocm
 build_relocatable=false
 build_address_sanitizer=false
+build_freorg_bkwdcomp=true
```
Comment (stanleytsang-amd) [inline install:44]: Shouldn't this be =false if the freorg-bkwdcomp command-line option turns it on?  Either that or have it the other way around...on by default and the command-line option turns it off...

File: install (hunk 1 above)
Comment (stanleytsang-amd) [reply install:117]: @TorreZuk I think you meant "If **on** by default..." but in any case I agree with Torre, it looks like its been coded to be on by default (see my comment on line 44 above) and the command-line option is also to turn it on...

File: CMakeLists.txt (hunk 2 above)
Comment (stanleytsang-amd) [reply CMakeLists.txt:115]: Agreed with @TorreZuk.  I also don't see FILE_REORG_BACKWARD_COMPATIBILITY being used anywhere within the actual library code.

File: header.hpp.in (hunk 3 above)
Comment (TorreZuk) [reply header.hpp.in:7]: Hmmm but on windows that is wrong, could just use include path /opt/rocm-ver/include/rocprim as well, right?

File: install (hunk 1 above)
Comment (arvindcheru) [reply install]: Updated flags in latest

File: CMakeLists.txt (hunk 2 above)
Comment (arvindcheru) [reply CMakeLists.txt]: Removed these unused setting in the latest

File: install (hunk 4 above)
Comment (arvindcheru) [reply install:44]: Updated flags in latest

**Review by TorreZuk — COMMENTED** [review]
Looks fine to me if it works, and the warning is more linux than windows specific.  Was this new structure affecting the window hipSDK organization as well?   Maybe CI will test this.
Leave it to Stanley to approve.

File: cmake/rocprim-backward-compat.cmake [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+	file(GLOB include_files ${PROJECT_FREORG_MAIN_INC_DIR}/*.hpp)
+	# Convert the list of files into #includes
+	foreach(include_file ${include_files})
+		get_filename_component(file_name ${include_file} NAME)
+		get_filename_component(file_name_we ${include_file} NAME_WE)
+		string(TOUPPER ${file_name_we} INC_FILE_NAME)
+		set(include_guard "${PROJECT_WRAPPER_MAIN_INC_GUARD}_${INC_FILE_NAME}_HPP")
+		set(include_statements "${PROJECT_WRAPPER_MAIN_INC_RELPATH}/${file_name}\"\n")
+		configure_file(${PROJECT_WRAPPER_TEMPLATE_FILE} "${PROJECT_WRAPPER_MAIN_INC_DIR}/${file_name}")
+		unset(INC_FILE_NAME)
```
Comment (lawruble13) [inline cmake/rocprim-backward-compat.cmake:89]: I don't think you need these `unset` statements, these are local variables which you're also setting immediately before they're used.

File: cmake/rocprim-backward-compat.cmake [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+set(PROJECT_WRAPPER_ITE_DET_GUARD       "${include_guard}ROCPRIM_WRAPPER_ITE_DET_INC")
+set(PROJECT_WRAPPER_WAR_DET_GUARD       "${include_guard}ROCPRIM_WRAPPER_WAR_DET_INC")
+
+
+
+# Function for Generating Wrapper Headers for Backward Compatibilty
+# Generates wrapper for all *.h files in include/rocprim folder
+# No Arguments to function.
+# Wrapper are generated under rocprim/include/rocprim.
+function (package_gen_bkwdcomp_hdrs)
```
Comment (lawruble13) [inline cmake/rocprim-backward-compat.cmake:78]: I think this function is massively overcomplicated, and could very easily be replaced with something simpler. For example,
```cmake
function(package_header_wrappers) # CMake function names usually avoid abbreviations
    file(GLOB_RECURSE include_files RELATIVE "${PROJECT_FREORG_MAIN_INC_DIR}" *.hpp)
    foreach(include_file ${include_files})
        get_filename_component(file_name ${include_file} NAME_WE)
        get_filename_component(file_path ${include_file} DIRECTORY)
        string(REPLACE '/' ';' path_dirs "${file_path}")
        set(rel_path "../../../include/rocprim")
        set(guard "ROCPRIM_WRAPPER")
        foreach(subdir IN_LIST path_dirs)
            if(NOT (subdir STREQUAL '' OR subdir STREQUAL '.'))
                set(rel_path "../${rel_path}/${subdir}")
                string(TOUPPER ${subdir} subdir_uc)
                set(guard "${guard}_${subdir_uc})
            endif()
        endforeach()
        string(TOUPPER ${file_name} file_name_uc)
        set(include_guard "${guard}_${file_name_uc}_HPP")
        set(include_statements "${rel_path}/${file_name}.hpp")
        configure_file(${PROJECT_WRAPPER_TEMPLATE_FILE} "${PROJECT_BINARY_DIR}/include/rocprim")
    endforeach()
endfunction()
```

File: cmake/rocprim-backward-compat.cmake [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+set(PROJECT_FREORG_INC_THREAD           ${PROJECT_SOURCE_DIR}/rocprim/include/rocprim/thread)
+set(PROJECT_FREORG_INC_WARP             ${PROJECT_SOURCE_DIR}/rocprim/include/rocprim/warp)
+set(PROJECT_FREORG_INC_BLK_DET          ${PROJECT_SOURCE_DIR}/rocprim/include/rocprim/block/detail)
+set(PROJECT_FREORG_INC_DEV_DET          ${PROJECT_SOURCE_DIR}/rocprim/include/rocprim/device/detail)
+set(PROJECT_FREORG_INC_DEV_SPE          ${PROJECT_SOURCE_DIR}/rocprim/include/rocprim/device/specialization)
+set(PROJECT_FREORG_INC_ITE_DET          ${PROJECT_SOURCE_DIR}/rocprim/include/rocprim/iterator/detail)
+set(PROJECT_FREORG_INC_WAR_DET          ${PROJECT_SOURCE_DIR}/rocprim/include/rocprim/warp/detail)
+
+#Relative Path for wrapper (include) files generated 
+set(PROJECT_WRAPPER_MAIN_INC_RELPATH    "${include_statements}#include \"../../../include/rocprim")
```
Comment (lawruble13) [inline cmake/rocprim-backward-compat.cmake:41]: The `include_statements` here aren't set. What is the point of having this variable in these strings?

File: header.hpp.in [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#ifndef @include_guard@
+#define @include_guard@ 
+/* The Wrapper template Configuration & settings
+ */
+#warning "This file is deprecated and will be removed in future releases. Use the include path /opt/rocm-ver/include and prefix with rocprim/"
+@include_statements@
```
Comment (lawruble13) [inline header.hpp.in:26]: Each of your files only includes one other header file, so you can do this instead and have `include_rel_file` set instead.
```suggestion
#include "@include_statements@"
```

**Review by lawruble13 — CHANGES_REQUESTED** [review]

File: rocprim/CMakeLists.txt [hunk 9]
```diff
@@ -53,27 +59,30 @@ set(INCLUDE_INSTALL_DIR ${CMAKE_INSTALL_INCLUDEDIR})
 # does not support header-only libraries (INTERFACE targets)
 rocm_install_targets(
   TARGETS rocprim rocprim_hip
-  PREFIX rocprim
 )
 
 rocm_install(
   DIRECTORY
     "include/"
     "${PROJECT_BINARY_DIR}/rocprim/include/"
```
Comment (lawruble13) [inline rocprim/CMakeLists.txt:67]: Why is this still `"${PROJECT_BINARY_DIR}/rocprim/include/"`?

**Review by saadrahim — CHANGES_REQUESTED** [review]
Changelog needs updating

File: header.hpp.in (hunk 8 above)
Comment (arvindcheru) [reply header.hpp.in]: comments updated on the latest version

File: cmake/rocprim-backward-compat.cmake (hunk 6 above)
Comment (arvindcheru) [reply cmake/rocprim-backward-compat.cmake]: Thanks for the comments/ref code shared.  updated on the latest version

File: cmake/rocprim-backward-compat.cmake (hunk 7 above)
Comment (arvindcheru) [reply cmake/rocprim-backward-compat.cmake]: Updated in the latest version

File: rocprim/CMakeLists.txt (hunk 9 above)
Comment (arvindcheru) [reply rocprim/CMakeLists.txt:67]: it was include only in the last version, Updated in the latest version with new structure

File: cmake/rocprim-backward-compat.cmake (hunk 5 above)
Comment (arvindcheru) [reply cmake/rocprim-backward-compat.cmake]: > I don't think you need these `unset` statements, these are local variables which you're also setting immediately before they're used.

Updated latest version with all review comments, kindly check thanks!!

**Comment by arvindcheru** [comment]
https://github.com/ROCmSoftwarePlatform/rocPRIM/pull/313 replaced with all changes (using new rocm-cmake wrapper approach) Hence closing

**Pull request closed without merging**